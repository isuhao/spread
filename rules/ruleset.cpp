#include "ruleset.hpp"
#include "urlrule.hpp"
#include <vector>
#include <map>
#include "misc/random.hpp"
#include <boost/thread/recursive_mutex.hpp>
#include <cstdio>

using namespace Spread;

typedef boost::shared_ptr<URLRule> URLPtr;
typedef boost::shared_ptr<ArcRuleData> ArcPtr;
typedef std::vector<URLPtr> UVec;
typedef std::map<Hash, UVec> UMap;
typedef std::map<Hash, ArcPtr> AMap;

#define LOCK boost::lock_guard<boost::recursive_mutex> lock(ptr->mutex)

struct RuleSet::_RuleSetInternal
{
  UMap urls;
  AMap arcs;

  Misc::Random rnd;
  boost::recursive_mutex mutex;

  const ArcRuleData* findArc(const Hash &hash) const
  {
    AMap::const_iterator it = arcs.find(hash);
    if(it == arcs.end())
      return NULL;

    return it->second.get();
  }

  // Adds all URL rules for 'hash' to the given vector
  void addURLs(const Hash &hash, RuleList &output) const
  {
    UMap::const_iterator it = urls.find(hash);
    if(it == urls.end()) return;

    const UVec &vec = it->second;
    if(!vec.size()) return;

    for(int i=0; i<vec.size(); i++)
      {
        const URLRule *r = vec[i].get();

        // Skip disabled rules
        if(r->isBroken) continue;

        // Add it!
        output.insert(r);
      }
  }

  // Returns an URL rule or NULL
  const Rule* findURL(const Hash &hash)
  {
    UMap::const_iterator it = urls.find(hash);
    if(it == urls.end())
      return NULL;

    const UVec &vec = it->second;

    if(!vec.size())
      return NULL;

    // First, build a list of rules with the highest priority
    int prio;
    bool prioFirst = true;
    std::vector<int> indices;
    double sum = 0;

    for(int i=0; i<vec.size(); i++)
      {
        const URLRule &r = *vec[i].get();

        // Skip disabled rules
        if(r.isBroken) continue;

        if(prioFirst)
          {
            prio = r.priority;
            prioFirst = false;
          }

        // Drop all previous results if this has a higher priority
        else if(r.priority > prio)
          {
            indices.clear();
            prio = r.priority;
            sum = 0;
          }

        // Skip everything below the highest priority level
        else if(r.priority < prio)
          continue;

        // Add this to the weight list
        if(r.weight > 0) sum += r.weight;
        indices.push_back(i);
      }

    if(indices.size() == 0)
      return NULL;

    /* Pick a card, any card!

       Generates a randum number between 0 and 'sum'.
     */
    double pick = rnd.dgen()*sum;

    // Pick the right one
    double psum = 0;
    for(int i=0; i<indices.size(); i++)
      {
        URLPtr ptr = vec[indices[i]];
        float w = ptr->weight;
        if(w <= 0) continue;
        psum += w;

        if(pick < psum)
          return ptr.get();
      }

    /* In case our impeccable math skillz failed, just go for the
       first one. This is also simpler than trying to account for
       pathological corner cases such as NaN weights etc.
    */
    return vec[indices[0]].get();
  }
};

RuleSet::RuleSet()
{
  ptr.reset(new _RuleSetInternal);
}

void RuleSet::findAllRules(const Hash &hash, RuleList &output) const
{
  LOCK;
  ptr->addURLs(hash, output);
}

const std::vector<Hash>* RuleSet::findHints(const Hash &dirHash) const
{
  LOCK;
  // Currently not implemented
  return NULL;
}

const Rule *RuleSet::findRule(const Hash &hash) const
{
  LOCK;

  // Search URLs only
  const Rule *rule = ptr->findURL(hash);
  if(rule) return rule;

  return NULL;
}

const ArcRuleData *RuleSet::findArchive(const Hash &hash) const
{
  LOCK;
  return ptr->findArc(hash);
}

void RuleSet::addArchive(const Hash &hash, const Hash &dirHash,
                         std::string ruleString)
{
  if(ruleString == "")
    ruleString = "ARC " + hash.toString() + " " + dirHash.toString();

  LOCK;
  ArcPtr data(new ArcRuleData(hash, dirHash, ruleString));

  // Make the archive retrievable both through the archive hash and
  // the dir hash.
  ptr->arcs[hash] = data;
  ptr->arcs[dirHash] = data;
}

void RuleSet::addURL(const Hash &hash, const std::string &url,
                     int priority, float weight,
                     std::string ruleString)
{
  if(ruleString == "")
    {
      ruleString = "URL " + hash.toString() + " ";

      bool hasWeight = (weight != 1.0);

      char buf[40];
      if(priority != 1 || hasWeight)
        {
          std::snprintf(buf,40,"%d ", priority);
          ruleString += buf;
        }
      if(hasWeight)
        {
          std::snprintf(buf,40,"%f ", weight);
          ruleString += buf;
        }

      ruleString += url;
    }

  LOCK;
  ptr->urls[hash].push_back
    (URLPtr(new URLRule(hash, ruleString, url, priority, weight)));
}

void RuleSet::reportBrokenURL(const Hash &hash, const std::string &url)
{
  LOCK;

  // Find all matching rules and disable them
  UMap::iterator it = ptr->urls.find(hash);
  if(it != ptr->urls.end())
    {
      UVec &vec = it->second;

      for(int i=0; i<vec.size(); i++)
        {
          URLRule *r = vec[i].get();
          if(r->url == url)
            r->isBroken = true;
        }
    }

  // Invoke the callback to notify external systems about the broken
  // URL.
  if(callback) callback(hash, url);
}
