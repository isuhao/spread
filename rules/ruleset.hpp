#ifndef __SPREAD_RULESET_HPP_
#define __SPREAD_RULESET_HPP_

#include "rulefinder.hpp"
#include <boost/shared_ptr.hpp>
#include <boost/function.hpp>

/* RuleSet is an implementation of the abstract RuleFinder interface.

   The RuleSet holds the global part of the rule set in memory. It
   holds URL and archive rules, but does not index hashes within the
   archives themselves.
 */

namespace Spread
{
  // These are used for the Rule::type field on all produced rules
  enum RuleSetTypes
    {
      RST_None          = 0,    // No/unknown rule type
      RST_URL           = 1,    // URL rule
      RST_Archive       = 2     // Archive / unpack rule
    };

  /* Specific Rule sublclasses for the rule types known to RuleSet.
   */
  struct URLRule : Rule
  {
    std::string url;
    bool isBroken;

    /* URL priorities are used to force some urls to always be chosen
       before others. If there exists more than one URL source for a
       given hash, the one with the highest priority is always
       returned. The only way a lower priority URL may be returned is
       if all the ones above it have failed (reported as broken.)

       This allows setting up fallback URLs to be used in case of
       temporary server failures, for example.

       If multiple sources share the highest priority, then weighting
       is used.
     */
    int priority;

    /* Weighting is used to pick from multiple sources that have the
       same priority. All the weights for the given priority level are
       summed, and each source is given a probability of weight/sum.
       Then a source is picked at random using those probabilities.
     */
    float weight;

    URLRule(const Hash &hash, const std::string &rulestr,
            const std::string &_url, int prio = 1, float wght = 1.0)
      : Rule(RST_URL, rulestr), url(_url), isBroken(false),
        priority(prio), weight(wght) { addOut(hash); }
  };

  struct RuleSet : RuleFinder
  {
    RuleSet();

    /* Inherited from RuleFinder.

       NOTE: Returned pointers are only valid for the lifetime of the
       RuleSet instance. Objects are deleted when RuleSet destructs.
     */
    const Rule *findRule(const Hash &hash) const;
    const Directory *findDir(const Hash &hash) const { return NULL; }

    /* Add an URL rule.
    */
    void addURL(const Hash &hash, const std::string &url,
                int priority = 1, float weight = 1.0,
                std::string ruleString = "");

    /* Decode and add rule string.

       Known formats:

       "URL <hash> <url> [priority] [weight]"
         : URL rule. See URLRule for description of the priority and
           weight fields.

       "ARC <arc-hash> <dir-hash>"
         : Archive rule. Archive in object <arc-hash> unpacks into the
           objects listed in <dir-hash>.
     */
    void addRule(const std::string &ruleString) {} // Not implemented yet

    /* Report broken URL rule. This will remove the rule from all
       future rule searches.

       This will also invoke a callback (if set through
       setURLCallback) which you can use to report the issue back to
       the server, or do other actions to correct the problem.
     */
    void reportBrokenURL(const Hash &hash, const std::string &url);

    // Get URLRule pointer from a Rule pointer
    static const URLRule *getURL(const Rule *r);

    // Set callback to handle broken URLs
    typedef boost::function< void(const Hash &hash, const std::string &url) > CBFunc;
    void setURLCallback(CBFunc cb) { callback = cb; }

  private:
    struct _RuleSetInternal;
    boost::shared_ptr<_RuleSetInternal> ptr;
    CBFunc callback;
  };

}

#endif
