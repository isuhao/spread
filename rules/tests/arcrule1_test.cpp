#include <iostream>
#include "ruleset.hpp"

#include <map>

using namespace std;
using namespace Spread;

Hash hello("hello",5);
Hash dolly("dolly",5);
RuleSet rules;

std::string h(const Hash &hash)
{
  return hash.toString().substr(0,8);
}

void test(const Hash &hash)
{
  const ArcRuleData *ard = rules.findArchive(hash);
  if(ard)
    {
      cout << "Found: arc=" << h(ard->arcHash) << " dir=" << h(ard->dirHash)
           << "\nString: " << ard->ruleString << "\n";
    }
  else
    {
      cout << h(hash) << " NOT FOUND!\n";
    }
}

void test()
{
  test(hello);
  test(dolly);
  cout << endl;
}

int main()
{
  cout << "Shooting blanks:\n";
  test();
  cout << "Adding one archive:\n";
  rules.addArchive(hello, dolly);
  test();
  cout << "Adding another archive:\n";
  rules.addArchive(dolly, dolly, "DOLLEH!");
  test();

  return 0;
}
