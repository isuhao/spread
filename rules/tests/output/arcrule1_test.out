Shooting blanks:
LPJNul-w NOT FOUND!
z5uehbma NOT FOUND!

Adding one archive:
Found: arc=LPJNul-w dir=z5uehbma
String: ARC LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF z5uehbmaSdLx6w2YGCXGTwadrwpgxfwZ-HRb9ePSkJ4F
Found: arc=LPJNul-w dir=z5uehbma
String: ARC LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF z5uehbmaSdLx6w2YGCXGTwadrwpgxfwZ-HRb9ePSkJ4F

Adding another archive:
Found: arc=LPJNul-w dir=z5uehbma
String: ARC LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF z5uehbmaSdLx6w2YGCXGTwadrwpgxfwZ-HRb9ePSkJ4F
Found: arc=z5uehbma dir=z5uehbma
String: DOLLEH!

